fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyUtils.java

```diff
@@ -0,0 +1,85 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.fineract.portfolio.delinquency.service;
+
+import java.math.BigDecimal;
+import java.time.LocalDate;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.fineract.organisation.monetary.domain.MonetaryCurrency;
+import org.apache.fineract.portfolio.loanaccount.domain.Loan;
+import org.apache.fineract.portfolio.loanaccount.domain.LoanRepaymentScheduleInstallment;
+import org.apache.fineract.portfolio.loanaccount.domain.LoanTransaction;
+import org.apache.fineract.portfolio.loanaccount.domain.LoanTransactionToRepaymentScheduleMapping;
+import org.apache.fineract.portfolio.loanaccount.domain.LoanTransactionType;
+
+public final class DelinquencyUtils {
+
+    private DelinquencyUtils() {}
+
+    public static LocalDate getOverdueSinceDate(final Loan loan, final LocalDate businessDate, final Integer graceOnArrearAgeing) {
+        final MonetaryCurrency loanCurrency = loan.getCurrency();
+        LoanRepaymentScheduleInstallment loanRepaymentSchedule = null;
+        for (LoanRepaymentScheduleInstallment installment : loan.getRepaymentScheduleInstallments()) {
+            if (installment.getDueDate().isBefore(businessDate) && installment.isNotFullyPaidOff()) {
+                loanRepaymentSchedule = installment;
+                break;
+            }
+        }
+
+        LocalDate overdueSinceDate = null;
+        if (loanRepaymentSchedule != null) {
+            // Default Due date
+            overdueSinceDate = loanRepaymentSchedule.getDueDate();
+        }
+
+        Map<Integer, BigDecimal> installmentBalances = new HashMap<>();
+        for (LoanRepaymentScheduleInstallment installment : loan.getRepaymentScheduleInstallments()) {
+            installmentBalances.put(installment.getInstallmentNumber(), installment.getPrincipalCompleted(loanCurrency).getAmount());
+        }
+
+        // If there is some Loan Transaction Chargeback
+        final List<LoanTransaction> loanTransactions = loan.retrieveListOfTransactionsByType(LoanTransactionType.CHARGEBACK);
+        for (LoanTransaction loanTransaction : loanTransactions) {
```

Comment:
This can be simplified:
- You already knows which is the first unpaid instalment
- Put the total paid principal for that instalment into a temporary variable
- If there are chargeback transactions for that instalment (fetch those loantransactions through the repayment mapping or simply iterate through the loan transactions and check the from- & due date of instalment against the transaction date), as iterating through those transaction, deduct the transaction amount from this temporary variable. And check after the deduction it went to negative. If yes, that transaction is not fully paid yet, so the delinquent amount can be collected from that moment and the overdue since date can be set with the transaction date of that transaction.
BTW question: only chargeback transaction should increase the delinquent amount?

Reply:
Done